src/base/http/requestparser.cpp

```diff
@@ -202,7 +202,7 @@ bool RequestParser::parseRequestLine(const QString &line)
             const QByteArray valueComponent = midView(param, (eqCharPos + 1));
             const QString paramName = QString::fromUtf8(QByteArray::fromPercentEncoding(nameComponent).replace('+', ' '));
             const QByteArray paramValue = valueComponent.isNull()
-                ? ""
+                ? QByteArray("")
```

Comment:
This statement holds: `"" == QByteArray("")` on both msvc & gcc.

I don't think you really identified some bug in the request parser. This looks like more of a workaround for msvc (if it really is the culprit). And your commit message should reflect that.

Reply:
@Chocobo1 

> This statement holds: `"" == QByteArray("")` on both msvc & gcc.

Sure. But unlike in that simple case, the trouble here, if it's not a Qt bug, is _most likely_ faulty logic in implicit conversion rules of the ternary expression and how they interact with `const char *` and `QByteArray`.

> This looks like more of a workaround for msvc (if it really is the culprit). And your commit message should reflect that.

Though I'm leaning towards that possibility, I have not confirmed it. It could also be a Qt bug. I also haven't read the standard, so I don't know which compiler is in the right here. It could very well be that it's GCC that is in the wrong, but we never noticed because it resulted the behavior we wanted. So it would be wrong to say that MSVC is the culprit in the commit message. Feel free to dig to the bottom of this, if you want (frankly I have no interest in that).

Either way, whether it works out for us or not, using `cond ? "" : QByteArray(...)` relies on obscure/non-obvious implicit conversion rules that no one wants to reason about when reading code. I would consider that a bug/bug-prone.

Reply:
>Either way, whether it works out for us or not, using cond ? "" : QByteArray(...) relies on obscure/non-obvious implicit conversion rules that no one wants to reason about when reading code. I would consider that a bug/bug-prone.

Whatever you like... I already stated the requirements in my last post (the last sentence to be precise) if I'm going to approve this change.

Reply:
@Chocobo1 

https://docs.microsoft.com/en-us/cpp/cpp/conditional-operator-q?view=vs-2019

>  **Warning**
>
>If the types of the second and third operands are not identical, then complex type conversion rules, as specified in the C++ Standard, are invoked. These conversions may lead to unexpected behavior including construction and destruction of temporary objects. For this reason, we strongly advise you to either (1) avoid using user-defined types as operands with the conditional operator or (2) if you do use user-defined types, then explicitly cast each operand to a common type.

Reply:
>using cond ? "" : QByteArray(...) relies on obscure/non-obvious implicit conversion rules that no one wants to reason about when reading code. I would consider that a bug/bug-prone.

Seems, @FranciscoPombal is right here. I, of course, did not go into all the subtleties of the standard, but at the top it is clear that the type of logical (ternary) operator is some type to which both operands (the second and third) are implicitly convertible to (so it can be either `QByteArray` or `const char *` in this case since  `QByteArray` provides implicit conversions from/to `const char *`). I don't think it's a good idea to let it run its course so we can consider this code as containing some bug.
>This statement holds: "" == QByteArray("") on both msvc & gcc.

Yes. But problem isn't this (second) operand. It's the next (third) one.
MSVC uses `const char *` as the result type of this operator so if third operand contains null character in its data it will be truncated at this character position when it is converted to QByteArray again.

To prevent such "dangerous" cases we even should use QT_NO_CAST_FROM_BYTEARRAY, IMO. @FranciscoPombal, can you test it (only this flag defined without your current changes)?

Reply:
Anyway commit message like current one isn't so good. It's better to say what program will do (will do correctly) after commit is applied unless it's really hard to formulate it. Even in this case it's wrong to have it in form of "Fix some_component" (this will result in a history with many duplicate entries, which is not very useful/meaningful). It's better to use "Fix short_description_of_bug" in this case.

Reply:
>>This statement holds: "" == QByteArray("") on both msvc & gcc.
>
>Yes. But problem isn't this (second) operand. It's the next (third) one.

Thanks for the explanation. Now I acknowledge the problem now.

>It's better to use "Fix short_description_of_bug" in this case.

👍

Reply:
@glassez 

Indeed, I have confirmed that adding `QT_NO_CAST_FROM_BYTEARRAY` to the compile definitions is an alternative way of fixing the problem. However, I wouldn't change this PR to _only_ include that change instead of the current one. I think both changes are necessary:

- The ternary expression change makes the code more correct independently of compile definitions and won't leave anyone wondering if there is any special reason as to why different types are used.
- The compile definition change guards against future (and possibly existing!) similar bugs, in case someone forgets to use the same type in both branches of conditional expressions involving `QByteArray`.

I'll backport the compile definition change to https://github.com/qbittorrent/qBittorrent/pull/12746 also.

As for the commit message, I see your point about including more detail in the title. I'm leaning towards something like this (the body is almost identical except for a slight reword and mentioning the change regarding the compile definitions):

`Fix truncation when parsing HTTP request query`

```
Closes #13029.

Fixes an issue with truncation of a QByteArray at the first '\0' byte
when parsing HTTP request query strings.

Previously, the operands of the ternary expression were of
different types. Most likely this was leading to a conversion
of the result to some kind of '\0'-terminated string type somewhere
along the way, in turn causing its truncation at the first '\0' byte
once converted back to QByteArray.

For some reason this bug was only present on Windows (MSVC).

Also added QT_NO_CAST_FROM_BYTEARRAY to the compile
definitions to safeguard against similar existing or future bugs.
```

If you're in agreement, I'll add the compile definitions change and rebase as a single commit with this message.

Reply:
>However, I wouldn't change this PR to only include that change instead of the current one.

I agree. It is better to avoid possible ambiguities in ternary operators.

>I'll add the compile definitions change and rebase as a single commit 

I wouldn't combine them in one commit. `QT_NO_CAST_FROM_BYTEARRAY` has sense in itself, even in other cases. Since many functions that accept `const char *` expect null-terminated string (and we often use QByteArray to store raw data), we should be more attentive to conversion cases, so implicit conversion can be harmful.

Reply:
@glassez @Chocobo1 fixed.